### ISSUE TITLE
Include support for satellite images with multiple bands

### ISSUE THREAD

**Issue by mirandadam** [FLIF-hub/FLIF#312]
Hello @jonsneyers !

Please include multiple bands support in the format.

  Forgive me if I missed something obvious, but as far as I understood, FLIF only supports GRAYSCALE, RGB and RGBA modes. Even though multispectral satellite images where not one of your benchmarks, geographical maps and medical images were, and those could benefit from multi band support.

  Also I would just like to confirm whether "Support for embedded ICC color profiles, Exif and XMP metadata" means that you could embed the entirety of the GeoTIFF metadata into a FLIF file. Is this correct?

  Storing satellite images is a very data intensive endeavour. Large datasets like planet.com's Dove satellite imagery, Digital Globe's high resolution multispectral archive, Landsat's multispectral archive since 1973, Brazilian INPE's archive of CBERS data and mirrors from other satellites' archives are all great examples of large scale commercial and governmental applications that could generate huge interest. That would motivate contributions from extremely qualified developers from the aerospace industry and probably attract direct investment.

 These players also could mean a route for popularization of the format. I don't think their needs are in the spotlight right now, since the focus seems to be on the web and cloud storage of personal pictures. If FLIF becomes a reliable lossless option for storing these archives, the space savings could be achieved without encumbering the end user (or their geoprocessing tools, for that matter) with a new format, just by engineering a translation layer to provide a geotiff file (that's what I would do in my own system).

I am sure you are aware of many other applications for multiple band imagery. I believe that is a big deal.

Congratulations on the great work on the FLIF image format and on fostering collaboration around it!

Best,
--Daniel

**Comment by jonsneyers** [comment]
You can embed anything in a FLIF file, it's basically just a matter of deciding on a convention for the chunk name.

4-channels maximum is currently hardcoded in FLIF, and in theory this can be extended but there are practical considerations: e.g. correlation between the channels is exploited by having information from other channels in the MANIAC context model (e.g. the chroma channels 'know' the luma value and the amount of luma prediction miss). Extending that approach to N channels can become a performance problem if N is large.

I don't have any experience with multiple band imagery. What is the current approach for compressing them? How many bands do you typically have, and how much correlation is there between them? Can you just save them as N grayscale images, or N/4 RGBA images?

I suppose if correlation between bands is not that important, you can save a multi band image as a grayscale animation, with one frame per band, and have a custom metadata chunk to provide some header information about which frame represents what.

**Comment by mirandadam** [comment]
Hello @jonsneyers ! Thanks for the quick follow up.

I started writing an answer to you and it became way too long. Here is the gist of it:
"Can you just save them as N grayscale images, or N/4 RGBA images?" "if correlation between bands is not that important" - I suppose yes, Landsat 8 imagery is actually distributed one band per file, but then you miss on the opportunity of using the correlation for compression, which certainly exists.

A long answer will follow soon.

**Comment by mirandadam** [comment]
Here is the full answer, I'll try my best to provide useful information:

1. "... N channels can become a performance problem if N is large.": Usually we are talking about N in the order of 4 to 10 for high resolution satellite imagery. Here are some examples:
- RapidEye Imagery: five 12-bit channels in a single geotiff
- Landsat 8 OLI: nine 12-bit channels in separate geotiffs, of which eight can be grouped into a single file.
- Leica ADS80 sensor (not a satellite): four 12-bit channels (RGBNir), option to output to a single geotiff. 
- LISS-III sensor: four 7-bit channels in a single geotiff.
- CBERS 4 MUX: four 8-bit channels
Lower resolution imagery is more complicated:
- MODIS - 36 12-bit channels, 29 of which can be stacked into a single file
- JPSS-1 CrIS: 1305 channels
Scientific hyperspectral imagery may have thousands of bands.

I am unable to assess the impact that N between 4 and 10 and/or N between 4 and 4000 would have on FLIF's performance, but the ideal scenario would probably be not to hard code a limit for the number of bands and have the performance degrade gracefully as the number of bands increases.

2. "...correlation between the channels..." and "chroma channels 'know' the luma value and the amount of luma prediction miss": I do not have the research background to answer to that, but there is clear correlation between the channels of a 5 channel rapideye image, and the correlation is stronger the "closer" the bands are, as in this [hyperspectral cube](http://large.stanford.edu/courses/2015/ph240/islam1/) (sample files [here](http://aviris.jpl.nasa.gov/data/free_data.html)). The spectral profile becomes a curve that can be interpolated, as long as you assume that the bands are not shuffled. I do not have the research background to answer about the best way to exploit that correlation, but it is certainly there (I do have work experience and academic training in image processing, however).

3. "What is the current approach for compressing them?": I do not know the literature about this so I cannot vouch for the state of the art, but the actual approach being used in production GeoTIFFs is rather simple. GDAL, a very complete library for handling image formats, uses LZW, DEFLATE and ZIP with a "predictor" option which can be either "none", "horizontal differencing" and "floating point predictor" (had to do some digging in the [source](https://github.com/OSGeo/gdal/blob/trunk/gdal/frmts/gtiff/libtiff/tif_predict.c) for that). I might be wrong, but not only this does not explore correlation between channels but also disregards correlation in the "vertical" direction. I did a quick research for the JPEG 2000 lossless algorithms, it seems to explore the correlation between bands.

I have quickly found a wealth of articles on the subject of compressing multispectral images. After a quick look though them I have selected [A Comparison of Multispectral Satellite Sequence Compression Approaches] (http://biblio.telecom-paristech.fr/cgi-bin/download.cgi?id=5803)  as a good primer. I also liked [CNES Studies of On-Board Compression for Multispectral and Hyperspectral Images](https://www.melaneum.com/documents/CThiebaut-SPIE-6683-04.pdf).

There is also a 2015 PHD Thesis: Karhunen-Loève Transform based Lossless Hyperspectral Image Compression for Space Applications, but it is waaaay to long to give a quick look through

**Comment by jonsneyers** [comment]
Thanks for the information. I suppose one way to implement this while staying within the FLIF16 bitstream, is as an animation of N/4 frames, where N is the number of bands. The bands that are most correlated would go in the same frame; that way you get to take at least _some_ advantage of the correlation between bands. And then define a metadata chunk to store GeoTIFF metadata, and another metadata chunk to encode the mapping between the animation frames/channels and the different bands.

**Comment by mirandadam** [comment]
Hi, Jon! That solution does take advantage on the current implementation with minor changes.

In my opinion you would lose generality in the format, however. Multispectral would be a corner case which could conceivably become poorly supported over time. Without some hacks to the decoder it would prevent one from having an animation with multispectral frames, for instance. Also it would require some thought on how to implement tiles, even if they are even being considered for animations.

I am trying to find time to read your [ICIP2016 paper](http://flif.info/papers/FLIF_ICIP16.pdf) and check out the slides. Do you believe a generalization for multispectral imagery is viable? If it is, please let me know, I would like to help.

Best,
Daniel

**Comment by jonsneyers** [comment]
Yes, generalization to "native" >4 channels is possible and not that hard to do, but it's outside the FLIF16 bitstream then.

The main thing that makes it a bit tricky to do at the moment, is that channels 0-3 are RGBA (with an optional special meaning for the alpha channel if the pixel is fully transparent), while channel 4 is not visible to the outside but it is used in animation to encode the FrameLookback channel (where pixels can refer to previous frames). So channel 4 is already in use, and it has a special meaning.

But if you just drop the FrameLookback transformation, then adding more channels is just a matter of adding support for them in image.hpp (currently it has 5 planes hardcoded), defining the order in which to encode them, and defining the MANIAC properties for each plane.

**Comment by mirandadam** [comment]
I opened [image.hpp](https://github.com/FLIF-hub/FLIF/blob/master/src/image/image.hpp) and saw the hardcoded parts you refer to, I also checked the predictors' code and that's where I got lost. Those would have to be adapted for multispectral data. On top of that there is a bitstream format change.

That seems to be overly complex for a simple patch, which was what I was hoping for.

Maybe this should be tried as part of another project and matured before being proposed for inclusion. It will be ok for me if you want to close this issue.

I am brainstorming requirements for a general multispectral image format in a repository of mine. Maybe I'll have something in the future.
